### ISSUE TITLE
Easiest Way to Access Request Object While Manipulating Response

### ISSUE THREAD

**Issue by eyazici** [httptoolkit/mockttp#71]
Hello there,

I am trying to develop a simple intercepting proxy using mockttp that stores all request/response pairs to a persistance storage in JSON form. Basically, I am using a pass through handler that captures upstream response just before the end of request&response lifecycle (ie. `beforeResponse` callback). However, I haven't managed to access the HTTP request object from inside `beforeResponse` callback. My code snippet is below. What is the most appropriate way to achieve this?

Thank you.

```	
server.forAnyRequest().always().thenPassThrough({
	beforeResponse: (response) => {
		let res = JSON.stringify(response);
		let req = JSON.stringify(response.getRequest()); // How to obtain request data of this response?
		
		// @TODO: Persist request&response pair to the storage in a serial form.
		return {};
	}		
});

```

**Comment by pimterry** [comment]
Hi @eyazici! Right now, this isn't possible using `beforeResponse`, although I see how it could be useful.

If you're looking to just observe request & response data though, this is possible using the `server.on('request', (req) => ...)` and `server.on('response', (res) => ...)` events. These events allow observing requests and responses independent of the individual rules that they match, and in an asynchronous and read-only way (so there's no performance impact, but those callbacks cannot modify either the request or response en route).

Those events include a `.id` property, which is consistent across all events related to a single request, so you can connect events together to understand what was sent and received.

Does that work for you?

**Comment by pimterry** [comment]
Closing for now, but feel free to reply if you're still having issues.

**Comment by eyazici** [comment]
Hello @pimterry, thank you for your explanations and the response. I followed your suggested method and maintain a map structure using `request.id` to track request-response relation like this:

```
var requests = {};

server.forAnyRequest().matching((request) => {	
	requests[request.id] = request; // Save request object for later use.
}).thenPassThrough({	
	beforeResponse: (response) => {
		if(response.id in requests) {
			const request = requests[response.id]; // Load the request object related to this response.
			// Process request object here.
			delete requests[response.id]; // Remove processed request from the 'request map'.
		}
		
		return response;
	}		
});
```
However it would be great to access related request object directly from the response object, like `response.request`.

**Comment by pimterry** [comment]
Yes, I think changing the callbacks to add an extra argument (i.e. `beforeResponse: (response, request) => { ... }`) might make sense, and it would be easily backwards compatible.

If you're interested in this, a pull request to add that would be welcome. There's some more notes on things to be aware of in my [last comment](https://github.com/httptoolkit/mockttp/issues/80#issuecomment-1124995694) from the thread in #80.
